### PR TITLE
Add Okta authentication support

### PR DESCRIPTION
**Pull request by sdford**
Adds and authentication connector for Okta as an alternative to OneLogin.

### CODE REVIEW

File: src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+     * @param relayState Deep link to redirect user to after authentication
+     * @return Session login token
+     */
+    protected AuthResult authenticateUser(final String username, final String password,
+                                          final String relayState) {
+
+        try {
+            return this.authClient.authenticate(username, password, relayState);
+        } catch (IOException ioe) {
+            final String msg = String.format("failed to get user groups for reason: %s", ioe.getMessage());
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java]: Did you mean to use a different error message here?  Or is this right?

File: src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    protected String getDeviceName(final Factor factor) {
+
+        Preconditions.checkArgument(factor != null, "factor cannot be null.");
+
+        final String factorProvider = factor.getProvider();
+        if (StringUtils.equalsIgnoreCase(factorProvider, "Google")) {
+            return "Google Authenticator";
+        }
+
+        return StringUtils.capitalize(factorProvider.toLowerCase());
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java]: Why toLowerCase() and then capitalize?  Seems odd.  Mixed case seems nicer anyway, unless there is a reason to cap?

File: src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+     * @param factor - Okta MFA factor
+     * @return Device name
+     */
+    protected String getDeviceName(final Factor factor) {
+
+        Preconditions.checkArgument(factor != null, "factor cannot be null.");
+
+        final String factorProvider = factor.getProvider();
+        if (StringUtils.equalsIgnoreCase(factorProvider, "Google")) {
+            return "Google Authenticator";
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java]: Capitalization logic doesn't get applied due to early return.

File: README.md [hunk 4]
```diff
@@ -99,7 +99,16 @@ auth.connector.onelogin.client_secret | Yes      | The OneLogin API client secre
...
     
 **Assumption: The current implementation looks up group membership for a user via the member_of field on the getUserById API response.**
-    
+
+##### Okta Auth Connector
+
+property                              | required | notes
+------------------------------------- | -------- | ----------
+auth.connector.okta.api_key           | Yes      | The Okta API key
+auth.connector.okta.base_url          | Yes      | The Okta base url (e.g. `https://nike.okta.com` or `https://nike.oktapreview.com`)
```
Comment (tlisonbee) [inline README.md]: Nitpick: It is probably better not to use Nike in our examples in general.  Better might be https://example.okta.com or https://acme.okta.com or similar

File: src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthConnector.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+        final AuthData authData = new AuthData();
+        final AuthResponse authResponse = new AuthResponse().setData(authData);
+
+
+        if (StringUtils.equals(authResult.getStatus(), OktaAuthHelper.AUTHENTICATION_MFA_REQUIRED_STATUS)) {
+            authResponse.setStatus(AuthStatus.MFA_REQUIRED);
+            authData.setStateToken(authResult.getStateToken());
+
+            embeddedData.getFactors().forEach(f -> authData.getDevices().add(new AuthMfaDevice()
```
Comment (fieldju) [inline src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthConnector.java]: Its generally frowned at to use variables that are hard to understand like `f`. Its preferred to just use variables names that have meaning like `factor`

File: src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthConnector.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        Preconditions.checkNotNull(authData, "auth data cannot be null.");
+
+        final List<UserGroup> userGroups = oktaAuthHelper.getUserGroups(authData.getUserId());
+
+        final Set<String> groups = new HashSet<>();
+        if (userGroups == null) {
+            return groups;
+        }
+
+        userGroups.forEach(g -> groups.add(g.getProfile().getName()));
```
Comment (fieldju) [inline src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthConnector.java]: See my previous comment about meaningful variable names.

File: src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+     *
+     * @param userId User ID
+     * @return User groups
+     */
+    protected List<UserGroup> getUserGroups(final String userId) {
+
+        try {
+            return this.userApiClient.getUserGroups(userId);
+        } catch (IOException ioe) {
+            final String msg = String.format("failed to get user groups for reason: %s", ioe.getMessage());
```
Comment (fieldju) [inline src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java]: Please include the user id in the logged message so when things go wrong we have more information for debugging.

File: src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+     * @param relayState Deep link to redirect user to after authentication
+     * @return Session login token
+     */
+    protected AuthResult authenticateUser(final String username, final String password,
+                                          final String relayState) {
+
+        try {
+            return this.authClient.authenticate(username, password, relayState);
+        } catch (IOException ioe) {
+            final String msg = String.format("failed to authenticate user for reason: %s", ioe.getMessage());
```
Comment (fieldju) [inline src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java]: Include the username in this msg, for debugging.

File: src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+     */
+    protected List<UserGroup> getUserGroups(final String userId) {
+
+        try {
+            return this.userApiClient.getUserGroups(userId);
+        } catch (IOException ioe) {
+            final String msg = String.format("failed to get user groups for reason: %s", ioe.getMessage());
+
+            throw ApiException.newBuilder()
+                    .withApiErrors(DefaultApiError.GENERIC_BAD_REQUEST)
```
Comment (fieldju) [inline src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java]: This doesn't seem to be the appropriate response.

See https://github.com/Nike-Inc/cerberus-management-service/blob/master/src/main/java/com/nike/cerberus/auth/connector/onelogin/OneLoginAuthConnector.java#L241

File: src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    protected AuthResult authenticateUser(final String username, final String password,
+                                          final String relayState) {
+
+        try {
+            return this.authClient.authenticate(username, password, relayState);
+        } catch (IOException ioe) {
+            final String msg = String.format("failed to authenticate user for reason: %s", ioe.getMessage());
+
+            throw ApiException.newBuilder()
+                    .withApiErrors(DefaultApiError.GENERIC_BAD_REQUEST)
```
Comment (fieldju) [inline src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java]: This should be a 401 of some sort.

Comment (fieldju) [inline src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java]: AUTH_BAD_CREDENTIALS

File: src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+        Preconditions.checkArgument(authResult != null, "auth result cannot be null.");
+
+        final Map<String, Object> embedded = authResult.getEmbedded();
+
+        try {
+            final String embeddedJson = objectMapper.writeValueAsString(embedded);
+            return objectMapper.readValue(embeddedJson, EmbeddedAuthResponseDataV1.class);
+        } catch (IOException e) {
+            throw ApiException.newBuilder()
+                    .withApiErrors(DefaultApiError.SERVICE_UNAVAILABLE)
```
Comment (fieldju) [inline src/main/java/com/nike/cerberus/auth/connector/okta/OktaAuthHelper.java]: Why would you want to return a 503 here?
It seems like this should be a 500 based error.

**Comment by sdford** [comment]
@fieldju thanks, I updated the PR

**Pull request merged**